### PR TITLE
Prevent multiple invitations and existing invitations

### PR DESCRIPTION
**Pull request by andreasfrisch**
closes dfds/cloudplatform/issues/2307

# Additional Review Notes
Prevents adding invitations for user-capability combinations which already have active invitations.
When accepting an invitaiton, all active invitations for same user-capability combination are cancelled.
When joining a capability by any means, all invitations for that same user and capability are cancelled.

### CODE REVIEW

File: src/SelfService/Application/InvitationApplicationService.cs [hunk 1]
```diff
@@ -91,6 +91,11 @@ public async Task<Invitation> AcceptInvitation(InvitationId invitationId)
                 throw new EntityNotFoundException("Capability does not exist");
             }
 
+            if (await _membershipRepository.IsAlreadyMember(capabilityId, invitation.Invitee))
+            {
+                invitation.Cancel();
+            }
+
             invitation.Accept();
```
Comment (kralle333) [inline src/SelfService/Application/InvitationApplicationService.cs:99]: first you cancel and then you accept? it looks a bit strange

File: src/SelfService/Application/InvitationApplicationService.cs [hunk 2]
```diff
@@ -149,6 +166,29 @@ Capability capability
...
+                _logger.LogWarning(
+                    "User \"{Invitee}\" is already member of \"{Capability}\"",
+                    inviteeId,
+                    capability.Id
+                );
+                continue;
+            }
+            var existingInvitations = await _invitationRepository.GetAllWithPredicate(
+                x => x.Invitee == invitee && x.TargetId == capability.Id && x.Status == InvitationStatusOptions.Active
+            );
```
Comment (kralle333) [inline src/SelfService/Application/InvitationApplicationService.cs:181]: maybe extract this to a function, you are doing the same in AcceptInvitation, but existingInvitations are called similarInvitations.

File: src/SelfService/Application/MembershipApplicationService.cs [hunk 3]
```diff
@@ -36,10 +38,20 @@ IMembershipApplicationDomainService membershipApplicationDomainService
...
     private async Task CreateAndAddMembership(CapabilityId capabilityId, UserId userId)
     {
+        // Note: requires [TransactionalBoundary], which should be wrapped elsewhere
+        var invitation = await _invitationRepository.FindByPredicate(
+            x => x.Invitee == userId && x.TargetId == capabilityId
+        );
+        if (invitation != null)
+        {
+            invitation.Cancel();
+        }
```
Comment (kralle333) [inline src/SelfService/Application/MembershipApplicationService.cs:53]: Here we are looking for a single invitation, but in InvitationApplicationService we are using GetAllWithPredicate

File: src/SelfService/Application/MembershipApplicationService.cs [hunk 4]
```diff
@@ -97,14 +109,7 @@ public async Task<MembershipId> AcceptApplication(MembershipApplicationId applic
             throw EntityNotFoundException<Capability>.UsingId(membershipApplication.CapabilityId);
         }
 
-        var newMembership = Membership.CreateFor(
-            capabilityId: membershipApplication.CapabilityId,
-            userId: membershipApplication.Applicant,
-            createdAt: _systemTime.Now
-        );
-
         await CreateAndAddMembership(membershipApplication.CapabilityId, membershipApplication.Applicant);
```
Comment (kralle333) [inline src/SelfService/Application/MembershipApplicationService.cs:112]: No longer need the id?

**Review by kralle333 — CHANGES_REQUESTED** [review]
I only have some smaller comments regarding consistency. I would like you to write tests to ensure that it works as intended regarding cancelation of invitations etc.

File: src/SelfService/Application/InvitationApplicationService.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreasfrisch) [reply src/SelfService/Application/InvitationApplicationService.cs]: Fixed

File: src/SelfService/Application/InvitationApplicationService.cs [hunk 5]
```diff
@@ -101,6 +106,18 @@ public async Task<Invitation> AcceptInvitation(InvitationId invitationId)
             );
             await _membershipRepository.Add(membership);
 
+            // cancel all similar invitations
+            var similarInvitations = await _invitationRepository.GetAllWithPredicate(
```
[the remark this answers is not in the document]
Comment (andreasfrisch) [reply src/SelfService/Application/InvitationApplicationService.cs]: Actually there should be none, as we cannot (should not be possible to) have duplicates -- this is purely a backup-failsafe

File: src/SelfService/Application/MembershipApplicationService.cs (hunk 4 above)
Comment (andreasfrisch) [reply src/SelfService/Application/MembershipApplicationService.cs:110]: This id is returned but never actually used by anything. (I am quite sure)

Besides that id is not the id we store, as `CreateAndAddMembership` also generates a new object.

File: src/SelfService/Application/MembershipApplicationService.cs (hunk 3 above)
Comment (andreasfrisch) [reply src/SelfService/Application/MembershipApplicationService.cs]: Good catch; Of course it should be all -- and also only active ones.

Again we should never have more than one, but to be on the safe side

File: src/SelfService/Application/InvitationApplicationService.cs (hunk 2 above)
Comment (andreasfrisch) [reply src/SelfService/Application/InvitationApplicationService.cs]: I still like the naming (as it makes sense in the context),  but I have extracted to functions that may have some better names.

**Review by kralle333 — APPROVED** [review]
nice!

**Pull request merged**